064-gate/proposal.md

```diff
@@ -0,0 +1,90 @@
+# Summary
+
+This RFC proposes a new step: `gate`.
+
+# Motivation
+
+Concourse pipeline lacks function of conditional workflow. Some use cases are:
+
+1. A pipeline runs code coverage scan. If coverage is lower than 85%, then fail the
+build.
+ 
+2. A pipeline monitors GitLab merged MergeRequests to automatically generate releases, 
+but if a MR is labeled with "no-release", then silently discard the MR.
+
+For the first use case, there could be two solutions: 1) Add gating function to the
+resource type of code coverage scan. This solution is not generic. With this direction,
+all similar scan related resource types would have to add gating function; 2) Use a 
+generic resource type of gating. This solution is better, but using a resource to do 
+gating is kinda overhead.
+
+For the second use case, gating resource seems to not fit. Because if a resource needs
+to abort a build, only way is to fail the build. If a user just wants to silently abort
+the build, then gating resource won't work. So that only solution is to make the GitLab
+resource to support to filter by MR labels, but which will make GitLab heavier and heavier.
+
+# Proposal
+
+A native step `gate` may make things easier and better.
+
+For the above use case 1, `gate` can be used like:
+
+```yaml
+- put: sonar-scan
+
+- load_var: scan-result
+  file: sonar-scan/result.json
+
+- gate: code-coverage-gate
+  condition: ((.:scan-result.coverage_rate)) < 0.85
+
+```
```

Comment:
@vito I think one way to think of the difference here is about configuring logic to run as part of the ATC scheduler vs. running the logic inside containers. Needing to pull a container from a registry, then launching the container, for very quick / simple logic turns into a bottleneck very quickly. Being able to express simple logic on the pipeline level can make execution of the logic much faster.

On the other hand, the single run step you propose is "more Concourse-y", as it's very typical for Concourse to launch many copies of small containers for light purposes (this is, after all, what check containers are). But it also contributed to fundamental performance limitations in Concourse's design for a long time, e.g. "dumb" webhooks to trigger checks instead of webhooks whose request bodies could be interpreted.

An RFC like this, to me, seems to try to straddle the line to provide more performant conditional pipeline flows.

Reply:
I think @vito's idea would be that the container that runs the coverage scan would be what performs the threshold check, rather than spinning up a new container for the comparison (correct me if I'm wrong)